packages/smithy-aws-core/src/smithy_aws_core/user_agent.py

```diff
@@ -0,0 +1,269 @@
+#  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#  SPDX-License-Identifier: Apache-2.0
+#  pyright: reportMissingTypeStubs=false,reportUnknownMemberType=false
+
+import os
+import platform
+from string import ascii_letters, digits
+from typing import NamedTuple, Optional, Self, Union, List
+
+from smithy_http.aio.crt import HAS_CRT
+
+_USERAGENT_ALLOWED_CHARACTERS = ascii_letters + digits + "!$%&'*+-.^_`|~"
+_USERAGENT_ALLOWED_OS_NAMES = (
+    "windows",
+    "linux",
+    "macos",
+    "android",
+    "ios",
+    "watchos",
+    "tvos",
+    "other",
+)
+_USERAGENT_PLATFORM_NAME_MAPPINGS = {"darwin": "macos"}
+_USERAGENT_SDK_NAME = "aws-sdk-python"
+
+
+class UserAgentComponent(NamedTuple):
+    """Component of a User-Agent header string in the standard format.
+
+    Each component consists of a prefix, a name, and a value. In the string
+    representation these are combined in the format ``prefix/name#value``.
+
+    This class is considered private and is subject to abrupt breaking changes.
+    """
+
+    prefix: str
+    name: str
+    value: Optional[str] = None
+
+    def to_string(self):
```

Comment:
Why isn't this just `__str__`? A limitation of `NamedTuple`?

Reply:
Good call, switched to __str__